connect/src/spirc.rs

```diff
@@ -800,6 +883,40 @@ impl SpircTask {
                     let now = self.now_ms();
                     self.device.set_is_active(true);
                     self.device.set_became_active_at(now);
+                    self.player.emit_session_connected_event(
+                        self.session.connection_id(),
+                        self.session.username(),
+                    );
+                    self.player.emit_session_client_changed_event(
+                        self.session.client_id(),
+                        self.session.client_name(),
+                        self.session.client_brand_name(),
+                        self.session.client_model_name(),
+                    );
+
+                    self.player
+                        .emit_volume_changed_event(self.device.get_volume() as u16);
+
+                    self.player
+                        .emit_auto_play_changed_event(self.session.autoplay());
+
+                    self.player.emit_filter_explicit_content_changed_event(
+                        self.session.filter_explicit_content(),
+                    );
+
+                    self.player
+                        .emit_shuffle_changed_event(self.state.get_shuffle());
+
+                    self.player
+                        .emit_repeat_changed_event(self.state.get_repeat());
+                }
+
+                let context_uri = update.get_state().get_context_uri().to_owned();
+
+                // completely ignore local playback.
+                if context_uri.starts_with("spotify:local-files") {
+                    self.notify(None)?;
```

Comment:
Do you want to return this potential error, or discard the result and always return `UnsupportedLocalPlayback` on the next line? What would make sense the most?

Reply:
`self.notify(None)?;` isn't returning an error? Notify just tells Spotify what our current state is (not playing the local track in this case) We are discarding the result by returning the error on the next line. If we don't notify before we discard the result Spotify thinks we're playing the local track and will show it as such in the client's UI. It leads to an inconsistent state between us and the client.

Reply:
We should prob also remove:

```rust
                    repeated.push(::std::convert::Into::into("audio/local"));
                    ...                  
                    repeated.push(::std::convert::Into::into("local"));
                    
```

Not sure we're advertising the ability to play local content if we can't in fact play local content?

Reply:
> `self.notify(None)?;` isn't returning an error?

But it can, it returns a `Result` after all. And the question mark operator is a shorthand to return from the function with that error, if any.

> Not sure why we're advertising the ability to play local content if we can't in fact play local content?

I agree. Unless someone knows how to add local content playback? Never understood that (nor did I try to).

Reply:
> But it can, it returns a Result after all. And the question mark operator is a shorthand to return from the function with that error, if any.

Well yes but it's not a return unless it's an error in which case we won't return the next line and the 1st error will be caught which is just as well, we'll have bigger problems on our hands? In any event we have to at least ignore local playback and not do the things in the following lines.

Reply:
No matter how you play it out notify has to be at least tried before you throw the error. Like I said if not Spotify thinks we're playing the local track.

Reply:
The alternative would be:
```rust
                if !context_uri.starts_with("spotify:local-files") {
                    if !self.state.get_track().is_empty() {
                        let start_playing =
                            update.get_state().get_status() == PlayStatus::kPlayStatusPlay;
                        self.load_track(start_playing, update.get_state().get_position_ms());
                    } else {
                        info!("No more tracks left in queue");
                        self.handle_stop();
                    }

                    self.notify(None)
                } else {
                    self.notify(None)?;
                    Err(SpircError::UnsupportedLocalPlayBack.into())
                }
```

Reply:
They are both functionally identical.

Reply:
> I agree. Unless someone knows how to add local content playback? Never understood that (nor did I try to).

You can't. The local files have to exist on the device AFAIK. In our case whatever librespot is running on. I don't think that any commercial Spotify connect device supports local playback?

Reply:
But I don't think that any client actually cares about ```kSupportedTypes``` I've removed those lines and the official clients still try to push local tracks. I think it's basically ignored and up to us what to do?

Reply:
> > But it can, it returns a Result after all. And the question mark operator is a shorthand to return from the function with that error, if any.
> 
> Well yes but it's not a return unless it's an error in which case we won't return the next line and the 1st error will be caught which is just as well, we'll have bigger problems on our hands? In any event we have to at least ignore local playback and not do the things in the following lines.

That’s my question. Suppose `notify` fails. Would you rather return that failure or still local playback unsupported? Which is dominant?

You could simply discard the result and still satisfy `clippy` with `let _ = self.notify()`

> But I don't think that any client actually cares about `kSupportedTypes` I've removed those lines and the official clients still try to push local tracks. I think it's basically ignored and up to us what to do?

Maybe but let’s do the right thing anyway and keep the things we support.

Reply:
> That’s my question. Suppose notify fails. Would you rather return that failure or still local playback unsupported? Which is dominant?

I would say that notify failing is a bigger problem.

> Maybe but let’s do the right thing anyway and keep the things we support.

So remove the `local` bits?

Reply:
OK. And yes let’s remove the local bits.

Reply:
> OK. And yes let’s remove the local bits.

Done